Job persistence

benarmston opened the pull request:
This PR adds persistence of the job registry and allocations registry to the controller.

`Allocation` and `Job` instances have learned how to serialize themselves and the classes have learned how to create new instances from those serializations.

A new class `Persistence` has been developed which saves and the serializations it is given and loads the serializations from the save file.  The save routines have been written to ensure that saving either the `JobRegistry` or `AllocationRegistry` is atomic.  This guarantee is the primary reason why the *registry* is saved rather than the individual jobs or allocations.

Various events in `EventProcessor` have been adjusted to save the registries when they are likely to have changed.

The `Allocation` class has changed so that it no longer has a reference to `Node`s, but rather stores the node name and then looks up the node from the `NodeRegistry`.  This makes (de)serialising an allocation simpler.

The environment for batch scripts and job steps are stored along side the job script.  The format is a set of `key=value` pairs separated by null chars.  Null chars have been chosen to allow the values to contain new lines.

Much of the serialization code could probably been cleaned up by using `ActiveModel::Attributes` or `ActiveModel::AttributeMethods`.  This has been left for a future task.

When the controller starts, prior to the Rack app being initialized, the job registry and allocation registry are loaded from disk.  Only once, this is complete does the Rack app start.  This results in the load being in a different thread/fibre than the main `Async` loop.  It turns out that `Enumerator`s do not support being called from different fibres/threads, the raise an exception if this is done.

To work around this, `Enumerator` has been removed from `ArrayTaskGenerator` and replaced with an integer index into the array job's array range.

benarmston committed:
Serialize job registry

benarmston committed:
ArrayTaskGenerator supports usage across fibres

It turns out that `Enumerator` doesn't like being called from different
`Fibre`s and therefore `Thread`s.  It raises an exception when this is
done.

In order to deserialize jobs, we will need be able to load the job and
therefore initialize the task generator from one fibre (or thread) and
use it from another.

We could try to work around that by making sure that we load the jobs in
the same fibre as is used by Falcon.  However, having
`ArrayTaskGenerator` support calls across threads is safer, easier and
more general.

Advancing the task pointed to by `ArrayTaskGenerator` is not thread
safe.  Advancing the task still needs to be synchronized, but there is
no exception when called by different threads.

benarmston committed:
Be explicit about setting @timer to timer

benarmston committed:
Deserialize job registry when app starts

benarmston committed:
Persist allocation registry

benarmston committed:
Tidy up errant allocations

benarmston committed:
Don't remove ARRAY_JOBs with allocated tasks

benarmston committed:
Improve (de)serialization code

benarmston committed:
Serialize batch script environment

benarmston committed:
Save job step environment

benarmston committed:
Fix job cleanup

benarmston committed:
Better error debugging

WilliamMcCumstie commented on app/models/batch_script.rb:
```diff
@@ -69,30 +79,70 @@ def stderr_path
+      else
+        File.read(env_path)
+      end
+    Hash[serialized_env.split("\0").map { |pairs| pairs.split('=', 2) }]
```
Is there are particular reason why we are using the `\0` and `=` syntax for envs? I suspect that it is something to do with the `\n` character.

However [ActiveMode::Serializer::JSON](https://api.rubyonrails.org/classes/ActiveModel/Serializers/JSON.html) should be able to encode newline characters perfectly fine, and makes serialisation easier.

benarmston replied:
Probably not a good one.  It is pretty much what slurm uses and is what came to mind when thinking about how to do this.

benarmston replied:
I'm really not sure if this or JSON would be better.  I think you might be right, but I have a nagging feeling that JSON has issues with storing arbitrary strings which I can't shake.  It would probably be good to revisit this later.

WilliamMcCumstie replied:
JSON can't store binary data, hence why it occasionally needs to be used with `base64`. However it should be able to escape any arbitrary UNICODE string.

The current method wouldn't be able to encode binary data as it would probably contain `\0` characters.

benarmston replied:
I've added a card to trello and this can be looked at at another time.

WilliamMcCumstie commented on app/models/job.rb:
```diff
@@ -160,7 +163,78 @@ def next_step_id
+    end
+  end
+
+  def self.from_serialized_hash(hash)
```
This method looks fine however I have a couple of code organisation queries. Firstly this is a class method, however it has been interspersed within the instance methods. Normally it should be located at the top of the file.

Secondly the `Job` model is getting very long and includes multiple disparate concepts. Putting the "serialization" code in the centre of the `Job` model makes it harder to determine what a "job" actually is.

A nicer solution would be to split the "serializer" code into a `HasSerializerableHash` concern. This way all the related serialisation code can be kept together without obscuring the purpose of `Job`.

I have a couple of reservations about calling it a "Serializer" considering we already have a `JobSerializer` class. However I can't think of a better name for it.

benarmston replied:
Having class methods at the top of the file is common.  However, having related methods close together is also common.  I think you're right, at the top of the class is likely most common.

What the the multiple disparate concepts that `Job` is now includes?

What models would include `HasSerializerableHash`?  The code in `from_serialized_hash` and `serializable_hash` are specific to `Job`.  I'm far from convinced that a single purpose concern included in only a single class provides for a nicer code organization.

I can't think of a better name either and as `serializable_hash` is the name chosen by `ActiveModel::Serialization` it needs to stay.

WilliamMcCumstie replied:
Just a general comment the file is getting long. We are starting to get attribute definitions, validation definitions, validation functions, and actual library code intermingling with each other.

The other models are fine, as their files are short. It just a particular problem for `Job` model. The concern would be `job` specific, no need to a generalised module. However something that allows better separation of concerns, so its easier to track what bit does what.

WilliamMcCumstie commented on app/models/job_step.rb:
```diff
@@ -66,18 +86,58 @@ def add_execution(node)
+      else
+        File.read(env_path)
+      end
+    Hash[serialized_env.split("\0").map { |pairs| pairs.split('=', 2) }]
```
Ditto with the above comment

WilliamMcCumstie commented on app/models/job_step.rb:
```diff
@@ -66,18 +86,58 @@ def add_execution(node)
+    Async::IO::Threads.new.async do
+      FileUtils.mkdir_p(dirname)
+      serialized_env = env.map { |k, v| "#{k}=#{v}" }.join("\0")
+      File.write(env_path, serialized_env)
```
NOTE: I am writing the comment here, but the same applies to anywhere `File.write` is used within a thread

Something to consider is `File.write` is not generally a thread safe operation. If the `write` method is called in rapid succession, the order the threads run in would be undefined due to their preemptive scheduling.

E.g. Running the following would generally write a bunch of `c` characters, but it could also write `a` or `b`. I couldn't see any interlacing problems, but no guarantee that wouldn't happen.
```
require 'parallel'
Parallel.map(['a','b','c'], in_threads: 3) { |letter| File.write('/tmp/demo', letter * 1000000) }
```

This is likely me being paranoid about data loss, however I'm not familiar enough with the code to know how frequently `write` is called. I am slightly hesitant about relying on `task.wait` not triggering another `write`.

The best way I can see around this is something like:

```
task = Async.task do |t|
  @write_mutex.synchronize do
    notify = Async::IO::Notification.new
    Thread.new do
       ... stuff ...
    ensure
      notify.signal
    end
    notify.wait
  end
end
task.wait
```
This is broadly based on how [Async::IO::Threads](https://github.com/socketry/async-io/blob/6eaa28808df5033a1965d749508e79348cb3356f/lib/async/io/threads.rb#L32) works, however with an added mutex to ensure only one thread is running at a time.

PS: There can also be issues if another thread tries to preform a read/ restore before the write has finished (possible started?). However I don't believe this happens anywhere?

benarmston replied:
My intention with this piece of code was to make the write atomic.  So that we would never get half of the registry written to the file.  I hadn't considered making the writes ordered.  You're right, we ought to do that too.

I can't see how interleaved reads and writes could cause any issues.  The backup is created with a hardlink.  This is an atomic operation.  The new version of the file is created with a rename.  This is also an atomic operation.  In both cases they cannot interleave with a read in any way that could cause partial or inconsistent data to be read.

Old data could be read.  That is worked around by only loading the registries when the application loads, prior to it accepting any connections or its periodic timer being started.  If this changes, we may need to prevent old data from being read.

benarmston replied:
The above comment was actually talking about `Persistence` not `JobStep`.

Regarding `JobStep`, the code is safe because it is called only once.  That can be made clearer.

WilliamMcCumstie commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -105,6 +124,11 @@ def with_lock
   end
 
+  def persistence
+    @persistence ||= FlightScheduler::Persistence
```
This looks a bit odd to me as there can only be a single `persistence` object due to the shared file path. However by creating a new instance it looks like I can have multiple allocation registries persist (which obviously isn't the case).

As the persistence is unitary, it would be better suited on `FlightScheduler.app`.

WilliamMcCumstie replied:
Possible something like the following:

```
class FlightScheduler::Application
  def save_allocation_registry
    @allocation_persistence.save(allocations.map(&:serializable_hash))
  end
end
```

benarmston replied:
I'm not sure what's odd here or why there can only be a single `persistence` instance.  Previously, I had both methods create a new persistence object each time they were called.

```
  def save
    FlightScheduler::Persistence
      .new('job registry', 'job_state')
      .save(jobs.map(&:serializable_hash))
  end
```

This works just fine.  But I didn't like the code duplication.  Perhaps creating `@persistence` in the initializer is better?

WilliamMcCumstie commented on lib/flight_scheduler/persistence.rb:
```diff
@@ -0,0 +1,100 @@
+
+    if Async::Task.current?
+      Async::IO::Threads.new.async do
+        block.call
```
As above manipulating file paths (in particular `FileUtils.mv(tmpfile.path, @path)`) is not thread safe. If two `saves` occur in rapid succession with different data, then the order the threads execute in is undefined. This will likely need to be coupled with a semaphore.

benarmston replied:
There is now a mutex ordering the calls.

WilliamMcCumstie commented on lib/flight_scheduler/persistence.rb:
```diff
@@ -0,0 +1,100 @@
+      #    path.
+      FileUtils.mkdir_p(File.dirname(@path))
+      if File.exist?(@path)
+        FileUtils.cp_lr(@path, @old_path, remove_destination: true)
```
I have hesitations about using hard links because they don't work on all file-systems ~(particularly NFS)~. I see the need for a link, however we should try and use symbolic if possible.

**EDIT:** I stand corrected, hard-links can be created over `nfs` as long as both files are on the same partition. Don't know if it works in general, but `nfs-utils` version `2.4.3` worked fine.

I propose changing the algorithm to be:

1. Generate a new unix time-stamped path: `/base/file/path.1606841756`
2. Write the new content to the time-stamped path
3. Reform the symbolic link from the actual path to the time-stamp
4. Remove the old timestamp path

Timestamps are nice because it allows for a more advanced retention policies (e.g. keeping snapshots), however UUIDs or a montonic incrementor would work just fine.

benarmston replied:
The limitation that hard links are only supported if both files are on the same partition is not limited to NFS.  It is a fundamental limitation of hard links.  NFS mounts support hard links as long as the exported filesystem supports hard links.

Creating a hardlink is an atomic operation.  It ensures that the backup will be exactly a copy of the main file and not a partial copy.  I believe that deleting and creating symlinks is not atomic and that there is no mechanism to edit a symlink.  So using symlinks would potentially result in a situation where the main file doesn't exist temporarily; or there is the possibility that it is a partial copy of the registry.

I may be wrong that symlinks are unsuitable.  However, I'm certain that I'm right that hard links are suitable, a.k.a. they can be used to ensure that the creation of the backup and the main file is atomic.

Your concern about the writes not being ordered is valid, but that is not related to the use hard/soft links.

The backup is in the same directory as the main file.  So it *really* ought to be on the same partition too.

Using multiple backups might be a good idea, but we then introduce a new problem of data retention.  I think that consideration is for another time.

WilliamMcCumstie commented on lib/flight_scheduler/application.rb:
```diff
@@ -85,9 +85,16 @@ def init_periodic_processor
+            job_registry.save
+            job_registry.jobs_in_state(Job::TERMINAL_STATES).each do |job|
+              Async.logger.info("Removing allocation for job in terminal state: id=#{job.display_id} state=#{job.state}")
+              FlightScheduler::Deallocation::Job.new(job).call
```
Why is this step required? I assume it is about getting the job out of the `TERMINAL_STATES` and fully removed from the scheduler?

Does the daemon know it needs to respond `NODE_DEALLOCATED` to all of these messages, even if it has sent it previously?

benarmston replied:
It a daemon is being restarted when the initial deallocation is being ran it won't have previously sent a `NODE_DEALLOCATED` message.  If that happens, we can end up with a job being stuck in a terminal state with nodes still allocated to it.

That can hold up other jobs from running through the scheduler.  Currently, the only solution to this was for a user to issue (another) cancellation request.

If the daemon has already received the deallocation request and responded with `NODE_DEALLOCATED` nothing breaks, the daemon will simply ignore the request.

WilliamMcCumstie replied:
From reviewing the daemon, it looks like it responds to all `JOB_DEALLOCATED` requests with `NODE_DEALLOCATED` (if it can).

I  vaguely remember this now from when I implemented it. At the time I considered the double response on cancel a nuisance but otherwise harmless (as you said).

However as the controller now relies on the positive `NODE_DEALLOCATED` response, it is critical that the daemon continues to respond to `JOB_DEALLOCATED` requests it does not recognise.

WilliamMcCumstie commented on app/models/batch_script.rb:
```diff
@@ -69,30 +79,70 @@ def stderr_path
       # We don't want the content hanging around in memory.
       self.content = nil
+
+      serialized_env = env.map { |k, v| "#{k}=#{v}" }.join("\0")
```
Ditto `JSON` serializer

WilliamMcCumstie reviewed (changes requested):
I have a couple of comments through out, which broadly categorised as:

1. Is there any particular reason we are using a custom serializer for `env` hashes as opposed to `JSON`?
2. We need to be careful of the thread safety of the various `write` and `save` methods

In reference to thread safety, it looks like the `job_step`, `allocation_registry` and `job_registry` now all preform some form of save/write. On closer inspection, the `job_step` is only saved to disk on create and is probably safe.

However the registries are saved on various events and in the concurrent timer task. This could lead to threads trying to write data to disk in rapid succession. This in turn cause various issues due to the thread's preemptive scheduling. See comment for details.

benarmston committed:
Re-order Job methods

benarmston committed:
Make it clearer that env is written once

benarmston committed:
Order persistence writes; fix concurrent write issue

Creating the hard link can raise an exception if it is already a link to
that file.

benarmston committed:
Prevent another jobs are stuck in CANCELLING/RUNNING issue

benarmston committed:
Fix logging issue

benarmston committed:
Reduce log verbosity

benarmston committed:
Stepd/Batchd don't use daemons controller connection

benarmston committed:
Fix another cancellation issue

WilliamMcCumstie commented on app/websocket_app.rb:
```diff
@@ -312,6 +305,52 @@ def call(env)
+  def process_daemon_connection(node_name, connection, message)
+    update_node_attributes(node_name, message)
+    Async.logger.info("#{node_name.inspect} connected")
+    processor = MessageProcessor.new(node_name, connection)
```
At some point we should look at splitting the `MessageProcessor` for the three different types:

* Regular `CONNECTED`,
* Batch connections, and
* Stepd connections (possible the same as above)

benarmston replied:
Yes we should.  Trello card exists to track this.

WilliamMcCumstie commented on lib/flight_scheduler/event_processor.rb:
```diff
@@ -70,7 +70,11 @@ def allocate_resources_and_run_jobs
     Async.logger.debug("Allocated nodes") {
-      allocated_nodes = FlightScheduler.app.allocations.each.map { |a| a.nodes }.flatten.sort.uniq
+      allocated_nodes = FlightScheduler.app.allocations.each
+        .map { |a| a.nodes }
```
Generally I would align this with the first `.` however its only a stylistic/ re-ability change.

Also whilst I am being a :panda_face:, this is using an `each.map`. I know this pre-existed the change, but just saw it

benarmston replied:
The allocation registry doesn't include `Enumerable`.  If it did, methods such as `drop` and `drop_while` would be available.  Perhaps that would be fine, or perhaps not.  I've deliberately avoid thinking about that.  Instead I opted to require users to use `each.<enumerable_method>`.  This is syntactic vinegar, which should cause the programmer to stop and think about whether they ought to be doing that or not.  In this particular case, I decided that `each.map` was OK for a debug message.

WilliamMcCumstie commented on lib/flight_scheduler/job_registry.rb:
```diff
@@ -70,14 +70,14 @@ def remove_old_jobs
           # Its possible that we're still processing the request to deallocate
           # the job.  If that's the case, we'll clean this job up another time.
-          if job.allocated?
+          if job.allocated? || FlightScheduler.app.job_registry.tasks_for(job).any?(&:allocated?)
```
This is quite an esoteric conditional. I assume that if an `array_job` has been allocated?

If so, its probably best to roll it into `job.allocated?`. Something like:

```
def allocated?
  if type == 'ARRAY_JOB'
    FlightScheduler.app.job_registry.tasks_for(job).any?(&:allocated?)
  else
     ...
  end
end
```

benarmston replied:
I thought about doing that, but I decided against it.  I don't want to end up in a situation where `job.allocated? == true` but `job.allocation == nil`.  That seems worse than this conditional.  Also, `job.allocation` returning an array of allocations does not seem right.  Perhaps a `job.allocated?(include_tasks: true)` would work?

WilliamMcCumstie commented on lib/flight_scheduler/job_registry.rb:
```diff
@@ -116,6 +116,29 @@ def running_tasks_for(job)
+  end
+
+  def save
+    persistence.save(jobs.map(&:serializable_hash))
```
Probably best to wrap this in a `read_lock`

benarmston replied:
The `jobs` method is already wrapped in a `read_lock`.  If I add one here too, an exception will be raised.  I could operate on `@jobs` directly, but then I'd be duplicating `jobs` in this method.

WilliamMcCumstie reviewed (commented):
A few more comments, nothing major to change.

Most of them could probably be addressed at a latter date.

benarmston committed:
Simplify conditional

benarmston commented:
I'm going to declare this good enough.  Thanks for your review of this @WilliamMcCumstie .
